### PR TITLE
Restrict local open in constructor argument types

### PR DESCRIPTION
**Pull request by gasche**
Forbid `type t = Foo of M.(bar * baz)` which could be understood as two separate constructor arguments `M.bar` and `M.baz` or a single tuple type.

See https://github.com/ocaml/ocaml/pull/12044#issuecomment-1607178102 -- thanks are due to @hypenrf for the report and @let-def for the implementation.

### CODE REVIEW

File: boot/menhir/menhirLib.ml [hunk 1]
```diff
@@ -3792,5 +3745,5 @@ module MakeEngineTable (T : TableFormat.TABLES) = struct
 end
 end
 module StaticVersion = struct
-let require_20210419 = ()
+let require_20220210 = ()
```
Comment (Octachron) [inline boot/menhir/menhirLib.ml:3748]: If we are increasing the menhir version, shouldn't we move to the last released version?

File: parsing/parser.mly [hunk 2]
```diff
@@ -3162,11 +3162,13 @@ generalized_constructor_arguments:
...
                                   { ($2,Pcstr_tuple [],Some $4) }
 ;
-
-constructor_arguments:
-  | tys = inline_separated_nonempty_llist(STAR, atomic_type)
+ constructor_arguments:
+  | atomic_type_without_open
+    %prec below_HASH
+      { Pcstr_tuple [$1] }
+  | ty = atomic_type STAR tys = inline_separated_nonempty_llist(STAR, atomic_type)
```
Comment (Octachron) [inline parsing/parser.mly]: Remark: we could also disallow `Foo of M.( x ) * N.( y )` to make it less surprising than `Foo of M.(x)` is a syntax error.

File: boot/menhir/menhirLib.ml (hunk 1 above)
Comment (gasche) [reply boot/menhir/menhirLib.ml:3748]: In my book, any move to a more recent version is fine.

File: parsing/parser.mly (hunk 2 above)
Comment (gasche) [reply parsing/parser.mly]: Hmm, or maybe it would be possible to allow `Foo of M.(x)`. I will try that and come back to confirm that it is hell.

Comment (Octachron) [reply parsing/parser.mly]: I would argue that having `Foo of M.(x -> y)` allowed but not `Foo of M.( x * y)` would not be an improvement in term of the readability of the grammar.

**Comment by gasche** [comment]
On a suggestion by @Octachron, local open is now systematically disallowed in constructor arguments for consistency: the change proposed by @let-def would allow `Foo of M.(r) * M.(t)` but reject `Foo of M.(r)`, and now both are rejected.

**Comment by yallop** [comment]
Personally, I'm happy with allowing `type t = Foo of M.(bar * baz)` and interpreting it as a single tuple argument, since I think of `M.(...)` as enhanced parentheses.

**Comment by gasche** [comment]
I don't necessarily disagree. On the other hand, I see the following arguments in favor of restricting:

- If unsure, restrict more, it is easy to allow later.
- If people make the wrong guess that `of M.(r * s * t)` has several arguments -- what would be written `M.(r and s and t)` in revised syntax -- then it is very hard for them to realize their mistake, and they silently get a worse value representation. In the direction, if you are surprised by a syntax error but really want to use the local-open syntax as a single argument, then writing `(M.(r * s * t))` is ugly but low-pain.

**Comment by hyphenrf** [comment]
> Personally, I'm happy with allowing `type t = Foo of M.(bar * baz)` and interpreting it as a single tuple argument, since I think of `M.(...)` as enhanced parentheses.

if that's the case, how would you express multiple constructor args with a local open?

**Comment by gasche** [comment]
@hyphenrf note that there isn't currently a syntax to allow a local open around several arguments. (And I'm quite sure that it would be a bad idea to use `of M.(... * ... * ...)` for this.).

One can of course use

```ocaml
open struct
  open M
  type t = ...
end
```

at the level of one or several structure items.

**Comment by yallop** [comment]
> if that's the case, how would you express multiple constructor args with a local open?

I don't think it matters if it's not possible to express that with a single local open; there are lots of things local opens can't do.  I think it's more important to have a uniform rule, like "`M.(te)` can go anywhere a type expression can go".

**Comment by lpw25** [comment]
> like "M.(te) can go anywhere a type expression can go"

I would argue that rule is respected either way here, as the debate is really about whether the argument to a constructor is a single type expression or a sequence of type expressions separated by *. I think it is the latter hence why I think it should be an error, but the other position is also consistent.

**Comment by yallop** [comment]
Disallowing `T of M.(x) * M.(y)`, as the PR currently does, seems unambiguously inconsistent with the rule.

**Comment by gasche** [comment]
To recap, there are three choices:

1. trunk: allow `of M.(foo)` and `of M.(a) * M.(b)`  
    justification: `M.(ty)` is a single type expression (despite the potential for confusion for `M.(a * b)`)
2. the first iteration of this PR: disallow `of M.(foo)` but allow `of M.(a) * M.(b)`  
    justification: `of M.(a * b)` could be perceived as ambiguous, so we reject it as a special case
3. the second/current iteration of this PR: disallow both `of M.(foo)` and `of M.(a) * M.(b)`
   justification: to avoid any perceived ambiguity, constructor arguments cannot start with a toplevel local-open

I find (1) and (3) both conceptually simpler than (2), and the grammar is also simpler.

In all cases `of (M.(foo))` is allowed, because this is clearly unambiguous (it is an explicitly-disambiguated form). There is no corresponding explicit-disambiguation for the other interpretation. (I think `M.(Foo of ...)` would be the most natural syntactic choice.)

**Comment by gasche** [comment]
I think that it may now too late to improve on this, local open in types has shipped without it and we haven't managed to build a consensus to restrict it here. Closing.

**Pull request closed without merging**